optics-core/src/Optics/ReadOnly.hs

```diff
@@ -0,0 +1,46 @@
+module Optics.ReadOnly
+  ( ReadOnly(..)
+  ) where
+
+import Optics.Internal.Bi
+import Optics.Internal.Optic
+import Optics.Internal.Profunctor
+
+-- | Class for read-write optics that have their read-only counterparts.
+class ReadOnly k where
+  type ReadOnlyOptic k
+  -- | Turn read-write optic into its read-only counterpart.
+  readOnly :: Optic k is s t a b -> Optic' (ReadOnlyOptic k) is s a
```

Comment:
isn't this composition with  `to id :: Getter a a`?

Reply:
And if it is, then `readOnly :: Getter a b -> Getter a b` etc. should be valid thing to do.

Reply:
It's not, the type signature of `\o -> o . to id` is too restricted.

Reply:
Then it should be explained in the comment, as I find it hard to believe.

Reply:
You don't need to believe it, you can see for yourself.

```haskell
λ> :t \o -> to id % o
\o -> to id % o
  :: (Is l (Join A_Getter l), Is A_Getter (Join A_Getter l)) =>
     Optic l js v v a b -> Optic (Join A_Getter l) js v v a b
λ> :t \o -> o % to id
\o -> o % to id
  :: (Is k (Join k A_Getter), Is A_Getter (Join k A_Getter)) =>
     Optic k is s t b b -> Optic (Join k A_Getter) is s t b b
```

Reply:
Now I don't understand:

```haskell
Prelude Control.Lens> toListOf (getting traverse) "foo"
"foo"
```

How that works, as AFAICS, `b` is not solved, and is ambiguous?

---

Some use example where indices of `Optic k is s t a b` cannot vary would be illustrative.

---

Anyway, my comment was related to the fact that in `lens` you can do:

```haskell
Prelude Control.Lens> view (getting (to id)) 'x'
'x'
```

So we should have `ReadOnly A_Getter` etc. instances for already readonly optics too (which would be (some) coerce).

Reply:
> How that works, as AFAICS, b is not solved, and is ambiguous?
I'm not sure what you mean.

```haskell
λ> :t getting
getting
  :: ToReadOnly k =>
     Optic k is s t a b -> Optic' (ReadOnlyOptic k) is s a
λ> :t getting traversed
getting traversed :: Traversable t => Optic' A_Fold NoIx (t a) a
λ> :t getting itraversed
getting itraversed
  :: TraversableWithIndex i f => Optic' A_Fold (WithIx i) (f a) a
λ> :t getting (itraversed % itraversed)
getting (itraversed % itraversed)
  :: (TraversableWithIndex i1 f1, TraversableWithIndex i2 f2) =>
     Optic' A_Fold '[i1, i2] (f1 (f2 a)) a
```

`getting` coerces `Optic k is s t a b` to `Optic k is s s a a` (using `Bicontravariant`) so `t` and `b` are irrelevant, they are ignored.

Reply:
> So we should have ReadOnly A_Getter etc. instances for already readonly optics too (which would be (some) coerce).

Right, we can add instances for read-only optics that don't do anything to the optic.

Reply:
Will it at least be the case that `ReadOnlyOptic k` is always `Join A_Getter k`? If so, do we need the extra type family?